src/main/software/amazon/event/ruler/Task.java

```diff
@@ -76,10 +76,27 @@ boolean stepsRemain() {
     }
 
     List<Object> getMatchedRules() {
-        return new ArrayList<>(rules);
+        return new ArrayList<>(matchingSubRules.stream()
+                .map(r -> r.getRule())
+                .collect(Collectors.toSet()));
     }
 
-    void collectRules(final NameState nameState) {
-      rules.addAll(nameState.getRules());
+    void collectRules(final Set<Double> candidateSubRuleIds, final NameStateWithPattern nameStateWithPattern) {
+        Set<NameState.SubRule> terminalSubRules = nameStateWithPattern.getNameState().getTerminalSubRulesForPattern(
+                nameStateWithPattern.getPattern());
+        if (terminalSubRules == null) {
+            return;
+        }
+
+        // If no candidates, that means we're on the first step, so all sub-rules are candidates.
+        if (candidateSubRuleIds.isEmpty()) {
+            matchingSubRules.addAll(terminalSubRules);
+        } else {
+            for (NameState.SubRule subRule : terminalSubRules) {
```

Comment:
this is on matching path, perf bound to number of candidate rules and number of terminate sub rules, but I see your point, there might not have large quantity of sub rules so unless generating dedicated benchmark with large sub rules, there may not have visible perf diff.

Reply:
Right, it would seem to be a very specifically crafted set of rules that would lead to a large number of terminal sub-rules on a single path for us to iterate through here.